### DISCUSSION TITLE
Streams API support for WebSockets

### DISCUSSION THREAD

**Discussion by ggarber** [kartikk221/hyper-express#19] [General]
Is there support in hyper-express for nodejs streams API?

This is available in the `ws` module and one of the possible use cases is to be able to pause/resume reading from a websocket to apply some back-pressure to the sender under high load:
https://github.com/websockets/ws#use-the-nodejs-streams-api

Or maybe there is a different way with hyper-express to pause reading from the websocket and apply this type of backpressure from the server to the client?

**Comment by kartikk221** [comment]
Yes, this is actually almost finished and should be pushed in the next update.

**Comment by kartikk221** [comment]
Websocket streaming is now available in update 5.4.0 with the Websocket.stream() method and the abilitiy to pipe a stream to the Websocket.writable property. The syntax/flow implemented is very similar to Response streaming but be careful to not initiate two streaming operations at the same time. You must wait for the stream() Promise to resolve or your piped readable stream to emit the 'close' event before initiating another streaming operation.

**Reply by vpalmisano** [reply]
I think that the question raised by @ggarber was related to streaming the messages received by the server and rate limiting the flow of received messages. The ws library uses the `socket.pause()` (https://nodejs.org/api/net.html#socketpause) function to limit the received data. I have not found a similar function in the uWebsocket library.

**Reply by kartikk221** [reply]
Ohh I see. There doesn't seem to be any method to pause or halt the incoming data from a uWS.Websocket in uWebsockets.js so I don't think that would be possible at this time.

**Reply by vpalmisano** [reply]
Just found this issue in the uSockets repository: https://github.com/uNetworking/uSockets/issues/27
